core/registry.js

```diff
@@ -0,0 +1,177 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview This file is a universal registry that provides generic methods
+ *    for registering and unregistering different types of classes.
+ * @author aschmiedt@google.com (Abby Schmiedt)
+ */
+'use strict';
+
+goog.provide('Blockly.registry');
+
+
+/**
+ * A map of maps. With the keys being the type and name of the class we are
+ * registering and the value being the constructor function.
+ * Ex: {'field': {'field_angle': Blockly.FieldAngle}}
+ *
+ * @type {Object<string, Object<string, function(new:?)>>}
+ */
+Blockly.registry.typeMap_ = {};
+
+/**
+ * A name with the type of the element stored in the generic.
+ * @param {string} name The name of the registry type.
+ * @constructor
+ * @template T
+ */
+Blockly.registry.Type = function(name) {
+  /** @private {string} */
+  this.name_ = name;
+};
+
+/**
+ * Returns the name of the type.
+ * @return {string} The name.
+ * @override
+ */
+Blockly.registry.Type.prototype.toString = function() {
+  return this.name_;
+};
+
+/** @type {!Blockly.registry.Type<Blockly.blockRendering.Renderer>} */
```

Comment:
Are you doing it this way instead of with an enum so that it's easy for someone else to add a type without changing core?

Reply:
Using generics this way allows the closure compiler to infer the type of the class in `register` and `getClass`, when the `Blockly.registry.Type` wrapper is passed in. ie: 
If you do: 
```
Blockly.registry.getClass(Blockly.registry.Type.RENDERER, 'zelos');
```
Using generics the closure compiler knows that the return type of that method is going to be a `Blockly.blockRendering.Renderer` type. 
However, if you do instead: 
```
Blockly.registry.getClass('renderer', 'zelos')
```
It wont. Since we cast that type to string, the two methods yield the same value, it's just the first one lets us internally infer the typechecks and: 
- a, not have to cast everywhere
- b, ensure that a type within the registry is consistent across, ie: all registered field types 
inherit from `Blockly.Field`.

This method avoids us having to add ``registerRenderer`, `registerField``, etc wrapper functions, and instead use the single register method but still type check.

Unfortunately, the parameter args of the class we're instantiating are still not type checked.

More context here: https://github.com/alschmiedt/blockly/pull/1

Reply:
That's awesome. Gets around a lot of the problems we were discussing. Very nice. :+1:

Reply:
Yep, it takes a little to get your head around it when reading the code, but I think it's worth it in this case.